tikvrpc/interceptor.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2021 TiKV Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikvrpc
+
+import (
+	"context"
+	"sync/atomic"
+)
+
+// Interceptor is used to decorate the RPC requests to TiKV.
+//
+// The definition of an interceptor is: Given an InterceptorFunc, we will
+// get the decorated InterceptorFunc with additional logic before and after
+// the execution of the given InterceptorFunc.
+//
+// The decorated InterceptorFunc will be executed before and after the real
+// RPC request is initiated to TiKV.
+//
+// We can implement an Interceptor like this:
+// ```
+// func LogInterceptor(next InterceptorFunc) InterceptorFunc {
+//     return func(target string, req *Request) (*Response, error) {
+//         log.Println("before")
+//         resp, err := next(target, req)
+//         log.Println("after")
+//         return resp, err
+//     }
+// }
+// txn.SetInterceptor(LogInterceptor)
+// ```
+//
+// Or you want to inject some dependent modules:
+// ```
+// func GetLogInterceptor(lg *log.Logger) Interceptor {
+//     return func(next InterceptorFunc) InterceptorFunc {
+//         return func(target string, req *Request) (*Response, error) {
+//             lg.Println("before")
+//             resp, err := next(target, req)
+//             lg.Println("after")
+//             return resp, err
+//         }
+//     }
+// }
+// txn.SetInterceptor(GetLogInterceptor())
+// ```
+type Interceptor func(next InterceptorFunc) InterceptorFunc
```

Comment:
It should be `RPCInterceptor`.

Reply:
I think there are two options:

1. Put the `Interceptor` in the `tikvrpc` package, the package name already contains the fact that it belongs to RPC.
2. Put the `Interceptor` in a separate package, then it's better to rename it to `RPCInterceptor`.

If we choose the second option, what package name do you think is more appropriate? At the moment I think it can be `tikvrpc/rpcinterceptor`, but it looks a bit verbose.

Reply:
how about `tikvrpc/interceptor`? It is distinguishable from the package import path.

Reply:
done